lazy-pages/src/host_func.rs

```diff
@@ -102,28 +102,31 @@ impl<'a> AccessHandler for HostFuncAccessHandler<'a> {
 fn accesses_pages(
     ctx: &mut LazyPagesExecutionContext,
     accesses: &[MemoryInterval],
-) -> Result<BTreeSet<GearPageNumber>, Error> {
-    let mut set = BTreeSet::new();
-    for access in accesses {
-        // TODO: here we suppose zero byte access like one byte access, because
-        // backend memory impl can access memory even in case access has size 0.
-        // We can optimize this if will ignore zero bytes access in core-backend (issue #2095).
-        let last_byte = access
-            .offset
-            .checked_add(access.size.saturating_sub(1))
-            .ok_or(Error::OutOfWasmMemoryAccess)?;
-
-        let page_size = GearPageNumber::size(ctx);
-        let mut offset = access.offset;
-        while offset <= last_byte {
-            set.insert(GearPageNumber::from_offset(ctx, offset));
-            offset = match offset.checked_add(page_size) {
-                Some(offset) => (offset / page_size) * page_size,
-                None => break,
+    pages: &mut BTreeSet<GearPageNumber>,
+) -> Result<(), Error> {
+    let page_size = GearPageNumber::size(ctx);
+
+    accesses
+        .iter()
+        .try_for_each(|access| -> Result<(), Error> {
+            // TODO: here we suppose zero byte access like one byte access, because
+            // backend memory impl can access memory even in case access has size 0.
+            // We can optimize this if will ignore zero bytes access in core-backend (issue #2095).
+            let last_byte = access
+                .offset
+                .checked_add(access.size.saturating_sub(1))
+                .ok_or_else(|| Error::OutOfWasmMemoryAccess)?;
+
+            let start = (access.offset / page_size) * page_size;
+            let end = (last_byte / page_size) * page_size;
+            let mut offset = start;
+            while offset <= end {
+                pages.insert(GearPageNumber::from_offset(ctx, offset));
+                offset += page_size;
```

Comment:
Where is checked add ? It is also very important  case

Reply:
Added